datastream-server/src/main/java/com/linkedin/datastream/server/assignment/BroadcastStrategy.java

```diff
@@ -59,17 +60,27 @@ public BroadcastStrategy(Optional<Integer> maxTasks) {
 
     Map<String, Set<DatastreamTask>> newAssignment = new HashMap<>();
 
+    Set<DatastreamTask> tasksAvailableToReuse = new HashSet<>();
     // Make a copy of the current assignment, since the strategy modifies it during calculation
     Map<String, Set<DatastreamTask>> currentAssignmentCopy = new HashMap<>(currentAssignment.size());
-    currentAssignment.forEach((k, v) -> currentAssignmentCopy.put(k, new HashSet<>(v)));
+    currentAssignment.forEach((k, v) -> {
+      currentAssignmentCopy.put(k, new HashSet<>(v));
+      // building the full list and then removing the reused tasks rather than building the list from dead tasks
+      // to avoid assignment of same task from dead instance to two live instances in case of interrupt exception.
+      tasksAvailableToReuse.addAll(v);
+    });
 
     for (String instance : instances) {
       newAssignment.put(instance, new HashSet<>());
       currentAssignmentCopy.putIfAbsent(instance, new HashSet<>());
+      tasksAvailableToReuse.removeAll(currentAssignmentCopy.get(instance));
     }
 
     int instancePos = 0;
     for (DatastreamGroup dg : datastreams) {
+      List<DatastreamTask> reuseTasksPerDg = tasksAvailableToReuse.stream().filter(x ->
+          x.getTaskPrefix().equals(dg.getTaskPrefix())).collect(Collectors.toList());
```

Comment:
Would it be better if we built a map of `taskPrefix` -> `List<DatastreamTask>` out of `tasksAvailableToReuse` before the loop on 80 instead of doing this? 

I think that'll be simpler, let alone it'll spare us having to filter `tasksAvailableToReuse` over and over for every `dg`.

Reply:
Done.